File: app/src/main/java/com/bopr/android/smailer/util/TagFormatter.java
package com.bopr.android.smailer.util;

import android.content.Context;
import android.content.res.Resources;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import androidx.annotation.NonNull;

/**
 * Tagged string formatter.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
public class TagFormatter {

    @SuppressWarnings("RegExpRedundantEscape") /* "\\}" works differently in android and pure java */
    private static final Pattern PATTERN = Pattern.compile("\\{(.*?)\\}");

    private final Map<String, String> values = new LinkedHashMap<>();
    private Resources resources;

    public static TagFormatter formatter(@NonNull Context context) {
        return new TagFormatter(context);
    }

    public static TagFormatter formatter(@NonNull Resources resources) {
        return new TagFormatter(resources);
    }

    public TagFormatter(@NonNull Resources resources) {
        this.resources = resources;
    }

    public TagFormatter(@NonNull Context context) {
        this(context.getResources());
    }

    public TagPattern pattern(String pattern) {
        return new TagPattern(pattern);
    }

    @Override
    @NonNull
    public String toString() {
        return "TagFormatter{" +
                "values=" + values +
                ", resources=" + resources +
                '}';
    }

    public TagPattern pattern(int resourceId) {
        return pattern(resources.getString(resourceId));
    }

    public class TagPattern {

        private String pattern;

        private TagPattern(String pattern) {
            this.pattern = pattern;
        }

        public TagPattern put(String key, String value) {
            values.put(key, value);
            return this;
        }

        public TagPattern put(String key, TagPattern value) {
            values.put(key, value.format());
            return this;
        }

        public TagPattern put(String key, int resourceId) {
            return put(key, resources.getString(resourceId));
        }

        public String format() {
            StringBuffer sb = new StringBuffer();

            Matcher matcher = PATTERN.matcher(pattern);
            while (matcher.find()) {
                String replacement = values.get(matcher.group(1));
                matcher.appendReplacement(sb, replacement != null ? replacement : "");
            }
            matcher.appendTail(sb);

            return sb.toString();
        }

        @Override
        @NonNull
        public String toString() {
            return format();
        }
    }
}

File: app/src/main/kotlin/com/bopr/android/smailer/util/TagFormatter.kt
package com.bopr.android.smailer.util

import android.content.Context
import android.content.res.Resources
import androidx.annotation.StringRes
import java.util.*
import java.util.regex.Pattern

/**
 * Tagged string formatter.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
class TagFormatter(private val resources: Resources) {

    private val values: MutableMap<String, String?> = LinkedHashMap()

    constructor(context: Context) : this(context.resources)

    fun pattern(pattern: String): TagPattern {
        return TagPattern(pattern)
    }

    fun pattern(@StringRes resourceId: Int): TagPattern {
        return pattern(resources.getString(resourceId))
    }

    override fun toString(): String {
        return "TagFormatter{" +
                "values=" + values +
                ", resources=" + resources +
                '}'
    }

    companion object {

        /* NOTE: "\\}" works differently on android device and pure java */
        @Suppress("RegExpRedundantEscape")
        private val VALUE_PATTERN = Pattern.compile("\\{(.*?)\\}")
    }

    inner class TagPattern internal constructor(private val pattern: String) {

        fun put(key: String, value: String?): TagPattern {
            values[key] = value
            return this
        }

        fun put(key: String, @StringRes resourceId: Int): TagPattern {
            return put(key, resources.getString(resourceId))
        }

        fun put(key: String, pattern: TagPattern): TagPattern {
            return put(key, pattern.format())
        }

        fun format(): String {
            val sb = StringBuffer()

            val matcher = VALUE_PATTERN.matcher(pattern)
            while (matcher.find()) {
                values[matcher.group(1)!!]?.let {
                    matcher.appendReplacement(sb, it)
                }
            }
            matcher.appendTail(sb)

            return sb.toString()
        }

        override fun toString(): String {
            return format()
        }

    }

}